### PR TITLE
(CDAP-1090) Feature/metrics search

### PR DESCRIPTION
**Pull request by gsps1**
Search for metrics context and available metrics.
https://issues.cask.co/browse/CDAP-1090

### CODE REVIEW

**Comment by abaranau** [comment]
>> We can’t automatically merge this pull request.

Please fix

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsDiscoveryQueryTestRun.java [hunk 1]
```diff
@@ -116,6 +119,55 @@ public void testMalformedPathReturns404() throws Exception {
     }
   }
 
+  @Test
+  public void testMetricsContexts() throws Exception {
+    metricsResponseCheck("/v2/metrics/available/context", 2);
```
Comment (abaranau) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsDiscoveryQueryTestRun.java]: don't you want to check actual values? not only number of those, right?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/MetricsEntityCodec.java [hunk 2]
```diff
@@ -97,17 +98,54 @@
...
     }
 
-    Arrays.fill(result, entityParts.length * idSize, depth * idSize, (byte) (padding & 0xff));
+    Arrays.fill(result, entityParts.length * idSize + offset, depth * idSize + offset, (byte) (padding & 0xff));
+  }
+
+  /**
+   * Encode with padding and return the rowkey. contextPrefix is passed as encoded byte-array while other
+   * parts are not encoded yet.
+   * @param contextPrefix
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/MetricsEntityCodec.java]: please describe params. not really straightforward

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsDiscoveryQueryTestRun.java [hunk 3]
```diff
@@ -116,6 +119,55 @@ public void testMalformedPathReturns404() throws Exception {
...
+    metricsResponseCheck("/v2/metrics/available/context/WordCount.f", 1);
+    metricsResponseCheck("/v2/metrics/available/context/WCount", 3);
+
+    String base = "/v2/metrics/available/context/WCount.f";
+    HttpResponse response = doGet(base);
+    Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+
+    String result = EntityUtils.toString(response.getEntity());
```
Comment (abaranau) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsDiscoveryQueryTestRun.java]: extract verify items in search result into separate method and reuse everywhere

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsDiscoveryQueryTestRun.java [hunk 4]
```diff
@@ -116,6 +119,55 @@ public void testMalformedPathReturns404() throws Exception {
...
+    List<String> resultList = new Gson().fromJson(result, new TypeToken<List<String>>() { }.getType());
+    Assert.assertEquals(2, resultList.size());
+    Assert.assertEquals("WCounter", resultList.get(0));
+    Assert.assertEquals("WordCounter", resultList.get(1));
+
+  }
+
+  private void metricsResponseCheck(String url, int expected) throws Exception {
```
Comment (abaranau) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsDiscoveryQueryTestRun.java]: on the comment above - add verification logic here

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 5]
```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+
+    return metricsScanResults;
+  }
+
+  /**
+   * Based on {@link co.cask.cdap.api.common.Bytes stopKeyForPrefix } but does the increment on the ByteBuffer instead.
+   * @param prefix
+   * @return
+   */
+  private static ByteBuffer stopKeyForPrefix(ByteBuffer prefix) {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: you better add another method in Bytes util class:

  public static byte[] stopKeyForPrefix(byte[] prefix, boolean changeInPlace) {...}

or better

  public static void stopKeyForPrefix(byte[] prefix) {...}

(returns void)

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 6]
```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+          metricsPart.position(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT));
+          metricsPart.limit(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT) +
+                              entityCodec.getEncodedSize(MetricsEntityType.METRIC));
+          metricsPart = stopKeyForPrefix(metricsPart);
+          if (metricsPart == null) {
+            scanner.close();
+            break;
+          }
+          //update the start-row to be row-key for next scan
+          startRow = rowKey;
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: shouldn't it be
startRow = metricsPart

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: oh, that's because you wrap it. Not easily readable. You are going to change wrapping (per other comment), so will be better

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 7]
```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+
+  /**
+   * Given the context and contextPrefix, this increments the context part after the contextPrefix, or the first context
+   * if no contextPrefix is provided. returns null if we cannot increment and have reached max value for the
+   * context part.
+   * @param context
+   * @param contextPrefix
+   * @return
+   */
+  private ByteBuffer getNextContextTarget(ByteBuffer context, String contextPrefix) {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: pass byte[], no need for ByteBuffer

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 8]
```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+    int idSize = entityCodec.getIdSize();
+    // if null , we only increment the first part of context, otherwise we increment the part after the contextPrefix.
+    int targetContextLocation = contextPrefix == null ? 0 : StringUtils.countMatches(contextPrefix, ".") + 1;
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: I don't think we use commons StringUtils anywhere.. Let's try not to bring another dependency.

Also, seems like you can calculate targetContextLocation once per query, no? Would make things more readable and efficient. Just create variable contextPartIndex or smth and use it inside while loop. Makes sense?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 9]
```diff
@@ -182,8 +325,8 @@ public void delete(String contextPrefix, String metricPrefix) throws OperationEx
     if (metricPrefix == null) {
       delete(contextPrefix);
     } else {
-      byte[] startRow = getPaddedKey(contextPrefix, "0", metricPrefix, null, 0, 0);
-      byte[] endRow = getPaddedKey(contextPrefix, "0", metricPrefix, null, Integer.MAX_VALUE, 0xff);
+      byte[] startRow = entityCodec.paddedEncode(contextPrefix, metricPrefix, null, 0, "0", 0);
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: what is this "0" ? RunId? Seems wrong, no?

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: shouldn't we pass null if we don't care?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 10]
```diff
@@ -369,20 +512,8 @@ private static void put(NavigableMap<byte[], NavigableMap<byte[], byte[]>> table
...
-      entityCodec.paddedEncode(MetricsEntityType.CONTEXT, contextPrefix, padding),
-      entityCodec.paddedEncode(MetricsEntityType.METRIC, metricPrefix, padding),
-      entityCodec.paddedEncode(MetricsEntityType.TAG, tagPrefix, padding),
-      Bytes.toBytes(timeBase),
-      entityCodec.paddedEncode(MetricsEntityType.RUN, runId, padding));
-  }
-
   private FuzzyRowFilter getFilter(MetricsScanQuery query, long startTimeBase,
-                                   long endTimeBase, boolean shouldMatchAllTags) {
+                                   long endTimeBase, boolean shouldMatchAllTags, byte[] timeBaseMatch) {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: why did you move it to param?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java [hunk 11]
```diff
@@ -140,6 +146,17 @@ public MetricsDiscoveryHandler(Authenticator authenticator, final MetricsTableFa
         return map;
       }
     });
+
+    this.timeSeriesTables = Suppliers.memoize(new Supplier<Map<MetricsScope, TimeSeriesTable>>() {
+      @Override
+      public Map<MetricsScope, TimeSeriesTable> get() {
+        Map<MetricsScope, TimeSeriesTable> map = Maps.newHashMap();
+        for (final MetricsScope scope : MetricsScope.values()) {
+          map.put(scope, metricsTableFactory.createTimeSeries(scope.name(), SECONDS_RESOLUTION));
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java]: so we always use highest resolution? Will it be more efficient if we use lowest? Or we may miss some data when searching that way?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java [hunk 12]
```diff
@@ -200,8 +217,82 @@ public void handleComponent(HttpRequest request, HttpResponder responder,
...
+    try {
+      responder.sendJson(HttpResponseStatus.OK, getNextMetrics(context));
+    } catch (OperationException e) {
+      LOG.warn("Exception while retrieving available metrics", e);
+      responder.sendStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
 
+  private List<String> getNextContext(String contextPrefix) throws OperationException {
+    List<String> nextLevelContexts = Lists.newArrayList();
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java]: Should it be a Set?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java [hunk 13]
```diff
@@ -200,8 +217,82 @@ public void handleComponent(HttpRequest request, HttpResponder responder,
...
+          String context = nextContext.substring(contextPrefix.length() + 1);
+          nextLevelContexts.add(context.substring(0, context.indexOf(".")));
+        }
+      }
+    }
+    Collections.sort(nextLevelContexts);
+    return nextLevelContexts;
+  }
+
+  private List<String> getNextMetrics(String contextPrefix) throws OperationException {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java]: getAvailableMetricNames?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 14]
```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+          //update the start-row to be row-key for next scan
+          startRow = rowKey;
+          metricsScanResults.add(contextStr);
+        } else {
+          // update the rowkey, by incrementing the metrics part
+          ByteBuffer metricsPart = ByteBuffer.wrap(rowKey);
+          metricsPart.position(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT));
+          metricsPart.limit(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT) +
+                              entityCodec.getEncodedSize(MetricsEntityType.METRIC));
+          metricsPart = stopKeyForPrefix(metricsPart);
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]:
```
// With the next scan we fast-forward to the next row that has different context or metric name part 
// that we are searching for.
startRow = getNextRow(rowKey, offset, length);

// Gets row key that is:
// * greater than given rowKey AND
// * has same bytes in the beginning (up to the given offset) AND
// * has different bytes in the range [offset..offset+length]
// or null if no row key with this properties exist.
byte[] getNextRow(rowKey, offset, length) {...}
```
in implementation do any simplest possible, e.g. use arrayCopy and Bytes.stopKeyForPrefix().

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: and compute offset and length before do {...} loop above as they don't change for the query.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 15]
```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+
+      rowResult = scanner.next();
+      if (rowResult != null) {
+        byte[] rowKey = rowResult.getRow();
+        // Decode context and metric from key
+        int offset = 0;
+        String contextStr = entityCodec.decode(MetricsEntityType.CONTEXT, rowKey, offset);
+
+        // Always have a "." suffix for unique matching
+        if (query.getContextPrefix() != null && !(contextStr + ".").startsWith(query.getContextPrefix())) {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: on the second thought you don't need + "."

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java (hunk 5 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: with the commend above - you don't need it

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java (hunk 8 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: after discussion: move this into a method in MetricsEntityCodec which holds this logic (see ENTITY_SPLITTER)

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java (hunk 10 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: after discussion: for no time range specified case one fuzzy rule is needed:
```
      fuzzyPairs.add(ImmutablePair.of(Bytes.concat(contextPair.getFirst(), metricPair.getFirst(), tagPair.getFirst(),
                                                   // fill time bytes with zeroes
                                                   FOUR_ZERO_BYTES, runIdPair.getFirst()),
                                      Bytes.concat(contextPair.getSecond(), metricPair.getSecond(), tagPair.getSecond(),
                                                   // FOUR_ONE_BYTES means anything in time bytes
                                                   FOUR_ONE_BYTES, runIdPair.getSecond())));
```
refactor

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java (hunk 11 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java]: after discussion: let's use lower resolution

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java [hunk 16]
```diff
@@ -55,14 +60,15 @@
 public final class MetricsDiscoveryHandler extends BaseMetricsHandler {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java]: please create JIRA to cleanup it after we remove 2.0 UI support and add todo linking to it here.

**Comment by abaranau** [comment]
couple comments. Let's finish it up today!

**Comment by abaranau** [comment]
confused with start/stop time in query - otherwise looks good

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 17]
```diff
@@ -148,6 +148,143 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+
+  /**
+   * If @param isContextQuery, is true we return the list of unique available contexts at the next level,
+   * if its false, we would return the available unique metrics in a given context
+   * @param query
+   * @param isContextQuery
+   * @return
+   * @throws OperationException
+   */
+  private List<String> getAvailableContextAndMetrics(MetricsScanQuery query, boolean isContextQuery)
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: This is a very long method. Please break it down.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 18]
```diff
@@ -148,6 +148,143 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+   * Returns null if no row key with this properties exist.
+   */
+  private byte[] getNextRow(byte[] rowKey, int offset, int length) {
+    byte[] stopKey = Bytes.stopKeyForPrefix(Arrays.copyOfRange(rowKey,
+                                                               offset, offset + length));
+    if (stopKey == null) {
+      return null;
+    }
+    // next row-key
+    return  Bytes.concat(Bytes.head(rowKey, offset), stopKey,
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: Still, this method is creating many `byte[]` (copyOfRange, stopKeyForPrefix, head, tail, concat). Anyway to reduce it?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java [hunk 19]
```diff
@@ -140,6 +150,17 @@ public MetricsDiscoveryHandler(Authenticator authenticator, final MetricsTableFa
         return map;
       }
     });
+
+    this.timeSeriesTables = Suppliers.memoize(new Supplier<Map<MetricsScope, TimeSeriesTable>>() {
+      @Override
+      public Map<MetricsScope, TimeSeriesTable> get() {
+        Map<MetricsScope, TimeSeriesTable> map = Maps.newHashMap();
+        for (final MetricsScope scope : MetricsScope.values()) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsDiscoveryHandler.java]: Remove final.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java (hunk 18 above)
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: @chtyim we are making an RPC call which is more expensive , so we decided to go with this.

Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: But still the head and tail are unncessary.

Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: improved that. please take a look. thanks

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java (hunk 17 above)
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: have split that down now.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java [hunk 20]
```diff
@@ -148,6 +148,143 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
...
+  private byte[] getNextRow(byte[] rowKey, int offset, int length) {
+    byte[] stopKey = Bytes.stopKeyForPrefix(Arrays.copyOfRange(rowKey,
+                                                               offset, offset + length));
+    if (stopKey == null) {
+      return null;
+    }
+    // next row-key
+    byte[] nextRow = new byte[rowKey.length];
+
+    copyByteArray(rowKey, nextRow, 0, 0 , offset);
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java]: a) allocate array of known size
b) use System.arrayCopy to fill prefix (that is unchanged)
c) use System.arrayCopy to fill stopKey in the rest

**Comment by abaranau** [comment]
please address the comment above and merge

**Comment by gsps1** [comment]
@abaranau addressed the comment and fixed a bug. please take a look to see if this is good, will merge after the build passes.

**Comment by abaranau** [comment]
LGTM
